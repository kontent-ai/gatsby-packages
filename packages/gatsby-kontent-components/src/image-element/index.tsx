import React from 'react';
import { GatsbyImageProps, GatsbyImage } from 'gatsby-plugin-image';
import {
  getGatsbyImageData,
  GetGatsbyImageDataProps,
} from './get-gatsby-image-data';

export interface ImageItem {
  image_id: string;
  url: string;
  description?: string;
  name?: string;
  height: number;
  width: number;
  type?: string;
}

export interface ImageElementProps
  extends GetGatsbyImageDataProps,
    Omit<GatsbyImageProps, 'image' | 'alt'> {
  alt?: string;
}

export const ImageElement: React.FC<ImageElementProps> = React.memo(function ImageElement({
  image,
  width,
  height,
  layout,
  backgroundColor,
  sizes,
  aspectRatio,
  options,
  alt,
  ...props
}): JSX.Element {
  const imageData = getGatsbyImageData({
    image,
    width,
    height,
    layout,
    backgroundColor,
    sizes,
    aspectRatio,
    options,
  });
  alt = alt ?? image.description ?? '';
  return <GatsbyImage image={imageData} {...props} alt={alt} />;
});
