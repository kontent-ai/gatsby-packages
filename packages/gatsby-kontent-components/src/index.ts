export { RichTextElement } from './rich-text-element';

export { ImageElement, getGatsbyImageData } from './image-element';
