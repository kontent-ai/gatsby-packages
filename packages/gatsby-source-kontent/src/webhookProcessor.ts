import { SourceNodesArgs, Node } from "gatsby"
import { CustomPluginOptions, KontentItem, KontentItemInput } from "./types"

import * as client from "./client";
import { addPreferredLanguageProperty, alterRichTextElements, getKontentItemLanguageVariantArtifact } from "./sourceNodes.items";
import { getKontentItemNodeStringForId, getKontentTaxonomyTypeName, getKontentTypeTypeName, RICH_TEXT_ELEMENT_TYPE_NAME, PREFERRED_LANGUAGE_IDENTIFIER, getKontentItemInterfaceName } from "./naming";
import { IWebhookDeliveryResponse, IWebhookMessage, IWebhookWorkflowResponse } from '@kentico/kontent-webhook-helper';
import _ from 'lodash';

const parseKontentWebhookBody = (api: SourceNodesArgs): IWebhookDeliveryResponse | IWebhookWorkflowResponse | null => {

  if ((api.webhookBody as IWebhookDeliveryResponse)?.message?.api_name === "delivery_preview" || (api.webhookBody as IWebhookDeliveryResponse)?.message?.api_name === "delivery_production") {
    const parsedBody = api.webhookBody as IWebhookDeliveryResponse;

    const isCorrectStructure = parsedBody?.data?.items?.every(item => item.language && item.id)
      && parsedBody?.message?.api_name
      && parsedBody?.message?.project_id
      && parsedBody?.message?.operation !== null;

    if (isCorrectStructure) {
      return parsedBody;
    }
  }
  else if ((api.webhookBody as IWebhookWorkflowResponse)?.message?.api_name === "content_management") {
    const parsedBody = api.webhookBody as IWebhookWorkflowResponse;

    const isCorrectStructure = parsedBody?.data?.items?.every(item => item.language && item.transition_from && item.transition_to && item.item.id)
      && parsedBody?.message?.api_name
      && parsedBody?.message?.project_id
      && parsedBody?.message?.operation !== null;

    if (isCorrectStructure) {
      return parsedBody;
    }
  }

  return null;
}

const isKontentSupportedWebhook = (message: IWebhookMessage, pluginConfig: CustomPluginOptions): boolean => {
  const isCorrectProject = message.project_id === pluginConfig.projectId;
  const isPreviewWebhook = 'delivery_preview' === message.api_name
    && ['upsert', 'archive', 'restore'].includes(message.operation);
  const isBuildWebhook = 'delivery_production' === message.api_name
    && ['publish', 'unpublish'].includes(message.operation);
  const isWorkflowWebhook = 'content_management' === message.api_name
    && message.operation === 'change_workflow_step';
  const isCorrectMessageType = message.type == 'content_item_variant';

  return isCorrectProject
    && (isPreviewWebhook || isBuildWebhook || (isWorkflowWebhook && pluginConfig?.experimental?.managementApiTriggersUpdate))
    && isCorrectMessageType
};

const createNodeFromRawKontentItem = (api: SourceNodesArgs, rawKontentItem: KontentItemInput, includeRawContent: boolean, preferredLanguage: string): string => {
  addPreferredLanguageProperty([rawKontentItem], preferredLanguage);
  alterRichTextElements([rawKontentItem]);
  const nodeData = getKontentItemLanguageVariantArtifact(
    api,
    rawKontentItem,
    includeRawContent,
  );
  api.actions.createNode(nodeData);
  return nodeData.id;
}

const isContentComponent = (data: KontentItem): boolean => {
  // Components have substring 01 in its id starting at position 14.
  // xxxxxxxx-xxxx-01xx-xxxx-xxxxxxxxxxxx
  const id = data?.system?.id;
  return id !== null && id.substring(14, 16) === "01";
}

const handleUpsertItem = async (
  api: SourceNodesArgs,
  pluginConfig: CustomPluginOptions,
  itemId: string,
  itemLanguage?: string,
): Promise<string[]> => {

  // TODO - language codename is not provided for management call
  if (itemLanguage && !pluginConfig.languageCodenames.includes(itemLanguage)) {
    api.reporter.verbose(`Cant find specified language ${itemLanguage} in plugin configuration`);
    return [];
  }

  // TODO could be optimized to by checking the fallback structure and save some requests
  // not recreate the ones that has different system.language
  // be careful on fallback language - verify cz->de->en fallbacks

  const createdItemsIds = [];
  for (const lang of pluginConfig.languageCodenames) {
    const { item: kontentItem, modularKontent } = await client.loadKontentItem(itemId, lang, pluginConfig, true);
    if (kontentItem === undefined) {
      api.reporter.verbose(`Kontent item (${itemId}) language variant (${lang}) not found on the kontent delivery API for update`);
      continue;
    }

    const nodeId = createNodeFromRawKontentItem(api, kontentItem, pluginConfig.includeRawContent, lang);
    createdItemsIds.push(nodeId);

    for (const key in modularKontent) {
      if (Object.prototype.hasOwnProperty.call(modularKontent, key)) {
        const modularKontentItem = modularKontent[key];
        const nodeId = createNodeFromRawKontentItem(api, modularKontentItem, pluginConfig.includeRawContent, lang);
        createdItemsIds.push(nodeId);
      }
    }
  }

  return createdItemsIds;
}

const handleDeleteItem = async (
  api: SourceNodesArgs,
  pluginConfig: CustomPluginOptions
): Promise<string[]> => {

  const itemInfo = (api.webhookBody as IWebhookDeliveryResponse)?.data.items[0];

  if (!pluginConfig.languageCodenames.includes(itemInfo.language)) {
    api.reporter.verbose(`Cant find specified language ${itemInfo.language} in plugin configuration`);
    return [];
  }

  // TODO could be optimized to by checking the fallback structure and save some requests
  // not recreate the ones that has different system.language
  // be careful on fallback language - verify cz->de->en fallbacks

  const touchedItemsIds = [];
  for (const lang of pluginConfig.languageCodenames) {
    const { item: kontentItem, modularKontent } = await client.loadKontentItem(itemInfo.id, lang, pluginConfig, true);
    if (kontentItem === undefined) { //item  was deleted (with content components)
      const idString = getKontentItemNodeStringForId(itemInfo.id, lang);
      const node = api.getNode(api.createNodeId(idString));

      // Remove content components
      const kontentItemNodes: KontentItem[] = api.getNodes()
        .filter((node: Node) => node.internal.type.startsWith(getKontentItemInterfaceName()))
        .map(node => node as KontentItem);

      const modularItemCodenames = _.flatMap(
        Object.values((node as KontentItem).elements)
          .filter(element => element.type === RICH_TEXT_ELEMENT_TYPE_NAME)
          .map(richTextElement => richTextElement.modular_content)
      );

      modularItemCodenames.forEach(modularItemCodename => {
        const candidate = kontentItemNodes.find((candidateNode) =>
          candidateNode?.system?.codename === modularItemCodename
          && candidateNode[PREFERRED_LANGUAGE_IDENTIFIER] === node[PREFERRED_LANGUAGE_IDENTIFIER])

        if (candidate && isContentComponent(candidate)) {
          touchedItemsIds.push(candidate.id);
          api.actions.deleteNode(candidate);
        }
      })


      if (node) {
        touchedItemsIds.push(node.id);
        api.actions.deleteNode(node);
      }
      continue;
    } else { // fallback version still available
      const nodeId = createNodeFromRawKontentItem(api, kontentItem, pluginConfig.includeRawContent, lang);
      touchedItemsIds.push(nodeId);

      for (const key in modularKontent) {
        if (Object.prototype.hasOwnProperty.call(modularKontent, key)) {
          const modularKontentItem = modularKontent[key];
          const nodeId = createNodeFromRawKontentItem(api, modularKontentItem, pluginConfig.includeRawContent, lang);
          touchedItemsIds.push(nodeId);
        }
      }
    }
  }

  if ((api.webhookBody as IWebhookDeliveryResponse)?.data.items.length > 1) {
    api.reporter.verbose(`Webhook contains more than one item for un-publish operation - upserting all (all but first in webhook) related items.`)
    for (const relatedItem of (api.webhookBody as IWebhookDeliveryResponse)?.data.items.slice(1)) {
      api.reporter.verbose(`Upserting item ${relatedItem.codename} (${relatedItem.codename}) - ${relatedItem.language}`)
      const processedIds = await handleUpsertItem(api, pluginConfig, relatedItem.id, relatedItem.language);
      touchedItemsIds.push(...processedIds);
    }
  }

  return touchedItemsIds;
}

const handleIncomingWebhook = async (
  api: SourceNodesArgs,
  pluginConfig: CustomPluginOptions,
  itemTypes: string[],
): Promise<void> => {

  const webhook = parseKontentWebhookBody(api);

  if (webhook === null) {
    api.reporter.verbose('Webhook ignored - webhook does not come from Kontent');
    return;
  }

  if (!isKontentSupportedWebhook(webhook.message, pluginConfig)) {
    api.reporter.verbose('This Kontent webhook is not handled by the Gatsby source kontent source plugin');
    return;
  }

  api.reporter.verbose(`Handling ${webhook.message.operation} from ${webhook.message.api_name} API`);
  if (webhook.data.items.length > 1) {
    if (webhook.message.operation != 'unpublish') {
      api.reporter.warn(`Webhook contains more than one item! - contains (${webhook.data.items.length})`)
    } else {
      api.reporter.verbose(`Unpublish webhook contains more than one item (${webhook.data.items.length})`)
    }
  }

  const processedItemIds: string[] = [];
  if (webhook.message.api_name === 'delivery_preview') {

    const item = (webhook as IWebhookDeliveryResponse).data.items[0];

    // TODO: Webhook header signature (once headers are available)
    // use signatureHelper '@kentico/kontent-webhook-helper'
    // https://github.com/gatsbyjs/gatsby/issues/23593

    if (webhook.message.operation === "upsert" || webhook.message.operation === "restore") {
      const processedIds = await handleUpsertItem(api, pluginConfig, item.id, item.language);
      processedItemIds.push(...processedIds);
    }

    if (webhook.message.operation === "archive") {
      const processedIds = await handleDeleteItem(api, pluginConfig);
      processedItemIds.push(...processedIds);
    }
  } else if (webhook.message.api_name === 'delivery_production') {

    const item = (webhook as IWebhookDeliveryResponse).data.items[0];

    // TODO: Webhook header signature (once headers are available)
    // use signatureHelper '@kentico/kontent-webhook-helper'
    // https://github.com/gatsbyjs/gatsby/issues/23593

    if (webhook.message.operation === "publish") {
      const processedIds = await handleUpsertItem(api, pluginConfig, item.id, item.language);
      processedItemIds.push(...processedIds);
    }

    if (webhook.message.operation === "unpublish") {
      const processedIds = await handleDeleteItem(api, pluginConfig);
      processedItemIds.push(...processedIds);
    }
  } else if (pluginConfig?.experimental?.managementApiTriggersUpdate && webhook.message.api_name === 'content_management') {

    const item = (webhook as IWebhookWorkflowResponse).data.items[0];

    // TODO: Webhook header signature (once headers are available)
    // use signatureHelper '@kentico/kontent-webhook-helper'
    // https://github.com/gatsbyjs/gatsby/issues/23593

    if (webhook.message.operation === "change_workflow_step") {
      const processedIds = await handleUpsertItem(api, pluginConfig, item.item.id);
      processedItemIds.push(...processedIds);
    } else {
      api.reporter.verbose(`Operation '${webhook.message.operation}' is not supported yet!`);
    }
  }
  else {
    api.reporter.verbose(`Webhook is not supported yet!`);
    api.reporter.verbose(JSON.stringify(webhook, null, 2));
    return;
  }

  for (const itemType of itemTypes) {
    const itemsToTouch = api.getNodesByType(itemType);
    itemsToTouch
      .filter(item => !processedItemIds.includes(item.id))
      .forEach(itemToTouch => api.actions.touchNode(itemToTouch))
  }

  if (pluginConfig.includeTaxonomies) {
    const taxonomies = api.getNodesByType(getKontentTaxonomyTypeName());
    for (const taxonomy of taxonomies) {
      api.actions.touchNode(taxonomy);
    }
  }

  if (pluginConfig.includeTypes) {
    const types = api.getNodesByType(getKontentTypeTypeName());
    for (const type of types) {
      api.actions.touchNode(type);
    }
  }
}

export {
  handleIncomingWebhook
}

