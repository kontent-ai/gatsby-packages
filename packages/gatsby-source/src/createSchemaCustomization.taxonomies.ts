import { CreateSchemaCustomizationArgs } from 'gatsby';
import { getKontentTaxonomiesSchemaNamingConfiguration } from './naming';

const createSchemaCustomization = async (
  api: CreateSchemaCustomizationArgs,
): Promise<void> => {
  const schema = getKontentTaxonomiesSchemaNamingConfiguration();
  api.actions.createTypes(schema);
};

export { createSchemaCustomization as kontentTaxonomiesCreateSchemaCustomization };
