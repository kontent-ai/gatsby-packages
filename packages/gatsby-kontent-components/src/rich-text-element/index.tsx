/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import parseHTML, { DomElement, domToReact } from 'html-react-parser';
import { ImageItem } from '../image-element';

const IMAGE_ID_ATTRIBUTE_IDENTIFIER = 'data-image-id';
const LINKED_ITEM_ID_ATTRIBUTE_IDENTIFIER = 'data-item-id';

interface Props {
  value: string;
  linkedItems?: any[];
  resolveLinkedItem?: Function;
  images?: ImageItem[];
  resolveImage?: ((image?: ImageItem) => JSX.Element) | undefined;
  links?: any[];
  resolveLink?: Function;
  resolveDomNode?: Function;
}

const isLinkedItem = (domNode: DomElement): boolean =>
  domNode.name === 'object' &&
  domNode.attribs?.type === 'application/kenticocloud';

const isImage = (domNode: DomElement): boolean =>
  domNode.name === 'figure' &&
  typeof domNode.attribs?.[IMAGE_ID_ATTRIBUTE_IDENTIFIER] !== 'undefined';

const isLink = (domNode: DomElement): boolean =>
  domNode.name === 'a' &&
  typeof domNode.attribs?.[LINKED_ITEM_ID_ATTRIBUTE_IDENTIFIER] !== 'undefined';

const replaceNode = (
  domNode: DomElement,
  linkedItems: any[] | undefined,
  resolveLinkedItem: Function | undefined,
  images: any[] | undefined,
  resolveImage: ((image?: ImageItem) => JSX.Element) | undefined,
  links: any[] | undefined,
  resolveLink: Function | undefined,
  resolveDomNode: Function | undefined,
): JSX.Element | void => {
  if (resolveLinkedItem && linkedItems) {
    if (isLinkedItem(domNode)) {
      const codeName = domNode.attribs?.['data-codename'];
      const linkedItem = linkedItems.find(
        item => item?.system?.codename === codeName,
      );
      return resolveLinkedItem(linkedItem,domNode);
    }
  }

  if (resolveImage && images) {
    if (isImage(domNode)) {
      const imageId = domNode.attribs?.[IMAGE_ID_ATTRIBUTE_IDENTIFIER];
      const image = images.find(image => image?.image_id === imageId);
      return resolveImage(image);
    }
  }

  if (resolveLink && links) {
    if (isLink(domNode)) {
      const linkId = domNode.attribs?.[LINKED_ITEM_ID_ATTRIBUTE_IDENTIFIER];
      const link = links.find(links => links?.link_id === linkId);
      return resolveLink(link, domNode);
    }
  }

  if (resolveDomNode) {
    return resolveDomNode(domNode, domToReact);
  }
};

const RichTextElement = ({
  value,
  linkedItems,
  resolveLinkedItem,
  images,
  resolveImage,
  links,
  resolveLink,
  resolveDomNode,
}: Props): JSX.Element => {
  const result = parseHTML(value, {
    replace: (domNode: DomElement) =>
      replaceNode(
        domNode,
        linkedItems,
        resolveLinkedItem,
        images,
        resolveImage,
        links,
        resolveLink,
        resolveDomNode,
      ),
  });

  if (result as JSX.Element[]) {
    return <>{result}</>;
  }

  return result as JSX.Element;
};

export { RichTextElement };
