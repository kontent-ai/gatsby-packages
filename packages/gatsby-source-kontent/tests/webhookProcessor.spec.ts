import * as _ from 'lodash';
import { SourceNodesArgs, Actions } from 'gatsby';
import { handleIncomingWebhook } from '../src/webhookProcessor';
import { CustomPluginOptions, KontentItem } from '../src/types';
import { createMock } from 'ts-auto-mock';
import { createContentDigest } from 'gatsby-core-utils';

import * as UPSERT_ITEM from './webhookUpsertClientItemResponse.json';
const UPSERT_ITEM_ID = UPSERT_ITEM.item.system.id;
const PROJECT_ID = "00676a8d-358c-0084-f2f2-33ed466c480a";
const LANGUAGE = "en-US";

import { mocked } from 'ts-jest/dist/util/testing';

jest.mock('../src/client', () => ({
  loadKontentItem: async (itemId: string): Promise<{
    item: KontentItem | undefined;
    modularKontent: { [key: string]: KontentItem };
  }> => {
    if (itemId.includes(UPSERT_ITEM_ID)) {
      return Promise.resolve(
        createMock<{
          item: KontentItem | undefined;
          modularKontent: { [key: string]: KontentItem };
        }>({
          ...UPSERT_ITEM
        })
      );
    } else {
      return Promise.resolve(
        createMock<{
          item: KontentItem | undefined;
          modularKontent: { [key: string]: KontentItem };
        }>({
          item: undefined,
          modularKontent: {}
        })
      );
    }
  }
}));

describe('webhookProcessor', () => {
  const pluginConfigurationBaseData = {
    projectId: PROJECT_ID,
    languageCodenames: [LANGUAGE]
  };
  const pluginConfiguration = createMock<CustomPluginOptions>(pluginConfigurationBaseData);

  describe('preview delivery API triggers', () => {
    describe('upsert action via upsert trigger', () => {

      const api = createMock<SourceNodesArgs>({
        webhookBody: {
          data: {
            items: [
              {
                id: UPSERT_ITEM_ID,
                codename: "author",
                language: LANGUAGE,
                type: "author"
              }
            ],
            taxonomies: []
          },
          message: {
            id: "3ad63b9a-b0c7-45ce-aa4f-a03ccee418ab",
            "project_id": PROJECT_ID,
            type: "content_item_variant",
            operation: "upsert",
            "api_name": "delivery_preview",
            "created_timestamp": "2020-09-22T08:43:06.9629422Z",
            "webhook_url": "https://testing-endpoint.io/__refresh"
          }
        },
        createNodeId: jest.fn(input => `dummyId-${input}`),
        actions: createMock<Actions>({
          createNode: jest.fn(),
        }),
        createContentDigest,
      });

      it('call createNode action for modular content as well', async () => {
        await handleIncomingWebhook(api, pluginConfiguration, []);

        const createNodesMock = mocked(api.actions.createNode, true);
        const createdNodes = _.flatMap(
          createNodesMock.mock.calls,
        ) as KontentItem[];
        expect(createNodesMock.mock.calls.length).toBe(2);
        expect(createdNodes).toMatchSnapshot();
      });
    });

    describe('delete operation via archive trigger', () => {

      const mainItemKontentSystemId = "74b9fc6e-b7e2-42a1-a01b-ae6ee69b2174";
      const inlineLinkedKontentItemCodename = "test_note_linekd_inline_kontent_item";
      const kontentComponentCodename = "bcbb41f3_b964_01af_7c70_8384a101bca5";

      const mainItemGraphQlNode = {
        "system": {
          "id": mainItemKontentSystemId,
          "name": "Ondrejch test (copy)",
          "codename": "ondrejch_test__copy_",
          "language": "en-US",
          "type": "author",
          "sitemap_locations": [],
          "last_modified": "2020-09-23T10:21:25.7632394Z"
        },
        "elements": {
          "notes": {
            "type": "rich_text",
            "name": "Notes",
            "images": [],
            "links": [],
            "modular_content": [
              inlineLinkedKontentItemCodename,
              kontentComponentCodename
            ],
            "value": `<p>Great notes</p>\n<object type="application/kenticocloud" data-type="item" data-rel="component" data-codename="${kontentComponentCodename}"></object>\n<object type="application/kenticocloud" data-type="item" data-rel="link" data-codename="${inlineLinkedKontentItemCodename}"></object>`
          }
        },
        "preferred_language": "en-US",
        "id": "c415f1ce-6022-5672-8439-04c023ee2184",
        "children": [],
        "internal": {
          "type": "kontent_item_author",
          "contentDigest": "f6d6bb19f908369cf2209a6de314000e",
          "counter": 144,
          "owner": "@kentico/gatsby-source-kontent"
        },
        "parent": null
      };

      const modularItemsGraphQlNodes = [
        {
          "system": {
            "id": "9616528a-d5f7-43ce-a313-1ff921adc078",
            "name": "test note linekd inline kontent item",
            "codename": "test_note_linekd_inline_kontent_item",
            "language": "en-US",
            "type": "note",
            "sitemap_locations": [],
            "last_modified": "2020-09-22T11:34:54.0552695Z"
          },
          "elements": {
            "text": {
              "type": "text",
              "name": "Text",
              "value": "test note linekd inline kontent item"
            }
          },
          "preferred_language": "en-US",
          "id": "ff8798ff-a586-5295-9a0b-3171194e2a47",
          "children": [],
          "internal": {
            "type": "kontent_item_note",
            "contentDigest": "a0b030791ddaee2291e024db0644e799",
            "counter": 268,
            "owner": "@kentico/gatsby-source-kontent"
          },
          "parent": null
        },
        {
          "system": {
            "id": "bcbb41f3-b964-01af-7c70-8384a101bca5",
            "name": "bcbb41f3-b964-01af-7c70-8384a101bca5",
            "codename": "bcbb41f3_b964_01af_7c70_8384a101bca5",
            "language": "en-US",
            "type": "note",
            "sitemap_locations": [],
            "last_modified": "2020-09-23T10:21:25.7632394Z"
          },
          "elements": {
            "text": {
              "type": "text",
              "name": "Text",
              "value": "Magic note"
            }
          },
          "preferred_language": "en-US",
          "id": "d02ab1e7-56d4-588d-9d15-0266a84ec349",
          "children": [],
          "internal": {
            "type": "kontent_item_note",
            "contentDigest": "aae62fcb99f01c62256c16f301df2eca",
            "counter": 145,
            "owner": "@kentico/gatsby-source-kontent"
          },
          "parent": null
        }
      ];

      const api = createMock<SourceNodesArgs>({
        webhookBody: {
          "data": {
            "items": [
              {
                "id": mainItemKontentSystemId,
                "codename": "ondrejch_test__copy_",
                "language": LANGUAGE,
                "type": "author"
              }
            ],
            "taxonomies": []
          },
          "message": {
            "id": "d9e52de0-e289-4275-bdd6-11ee00d0c799",
            "project_id": PROJECT_ID,
            "type": "content_item_variant",
            "operation": "archive",
            "api_name": "delivery_preview",
            "created_timestamp": "2020-09-22T11:44:31.7718735Z",
            "webhook_url": "https://testing-endpoint.io/__refresh"
          }
        },
        createNodeId: jest.fn(input => `dummyId-${input}`),
        getNodes: jest.fn(() => [mainItemGraphQlNode, ...modularItemsGraphQlNodes]),
        getNode: jest.fn(() => mainItemGraphQlNode),
        actions: createMock<Actions>({
          deleteNode: jest.fn(),
        }),
        createContentDigest,
      });

      it('call deleteNode action for modular content as well', async () => {
        await handleIncomingWebhook(api, pluginConfiguration, []);

        const deletedNodesMock = mocked(api.actions.deleteNode, true);
        const deletedNodes = _.flatMap(
          deletedNodesMock.mock.calls,
        ) as KontentItem[];
        expect(deletedNodesMock.mock.calls.length).toBe(2);
        expect(deletedNodes).toMatchSnapshot();
      });
    });
  });

  describe('publish delivery API triggers', () => {

  });
  describe('management API triggers', () => {
    describe('upsert action via change workflow step trigger (experimental)', () => {
      const api = createMock<SourceNodesArgs>({
        webhookBody: {
          data: {
            items: [
              {
                item: {
                  id: UPSERT_ITEM_ID
                },
                language: {
                  id: "00000000-0000-0000-0000-000000000000"
                },
                "transition_from": {
                  id: "88ac5e6e-1c5c-4638-96e1-0d61221ad5bf"
                },
                "transition_to": {
                  id: "10b99292-5bd1-4d4e-b6c1-a2fe2aedf3c5"
                }
              }
            ],
            taxonomies: []
          },
          message:
          {
            id: "562a9727-61b5-4919-9083-105b0c64995b",
            "project_id": PROJECT_ID,
            type: "content_item_variant",
            operation: "change_workflow_step",
            "api_name": "content_management",
            "created_timestamp": "2021-07-01T10:19:25.4983423Z",
            "webhook_url": "https://testing-endpoint.io/__refresh"
          }
        },
        createNodeId: jest.fn(input => `dummyId-${input}`),
        actions: createMock<Actions>({
          createNode: jest.fn(),
        }),
        createContentDigest,
      });


      it('does not call createNode actions when turned off', async () => {
        await handleIncomingWebhook(api, pluginConfiguration, []);

        const createNodesMock = mocked(api.actions.createNode, true);
        const createdNodes = _.flatMap(
          createNodesMock.mock.calls,
        ) as KontentItem[];
        expect(createNodesMock.mock.calls.length).toBe(0);
        expect(createdNodes).toMatchSnapshot();
      });

      it('call createNode action for modular content as well when turned on', async () => {

        const experimentalPluginConfiguration = createMock<CustomPluginOptions>({
          ...pluginConfigurationBaseData,
          experimental: {
            managementApiTriggersUpdate: true
          }
        });

        await handleIncomingWebhook(api, experimentalPluginConfiguration, []);

        const createNodesMock = mocked(api.actions.createNode, true);
        const createdNodes = _.flatMap(
          createNodesMock.mock.calls,
        ) as KontentItem[];
        expect(createNodesMock.mock.calls.length).toBe(2);
        expect(createdNodes).toMatchSnapshot();
      });

    });
  });




});
