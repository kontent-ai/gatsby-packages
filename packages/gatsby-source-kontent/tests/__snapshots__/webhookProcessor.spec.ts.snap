// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webhookProcessor management API triggers upsert action via change workflow step trigger (experimental) call createNode action for modular content as well when turned on 1`] = `
Array [
  Object {
    "children": Array [],
    "elements": Object {
      "notes": Object {
        "images": Array [],
        "links": Array [],
        "modular_content": Array [
          "n6951be2b_e502_01ed_79a4_80ad776ae16b",
        ],
        "name": "Notes",
        "type": "rich_text",
        "value": "<p>This is a sample notes element with one note component</p>
<object type=\\"application/kenticocloud\\" data-type=\\"item\\" data-rel=\\"component\\" data-codename=\\"n6951be2b_e502_01ed_79a4_80ad776ae16b\\"></object>",
      },
    },
    "id": "dummyId-kontent_item_en-US_eedf0209-989c-4adb-bc83-546e325ab781",
    "internal": Object {
      "contentDigest": "4416e4041857112eddf4caeb6dca5195",
      "type": "kontent_item_author",
    },
    "preferred_language": "en-US",
    "system": Object {
      "codename": "author",
      "id": "eedf0209-989c-4adb-bc83-546e325ab781",
      "language": "en-US",
      "last_modified": "2020-09-22T09:33:13.7465607Z",
      "name": "Author",
      "sitemap_locations": Array [],
      "type": "author",
    },
  },
  Object {
    "children": Array [],
    "elements": Object {
      "text": Object {
        "name": "Text",
        "type": "text",
        "value": "Note 1",
      },
    },
    "id": "dummyId-kontent_item_en-US_6951be2b-e502-01ed-79a4-80ad776ae16b",
    "internal": Object {
      "contentDigest": "5d088a503493a36e057d990043b1515e",
      "type": "kontent_item_note",
    },
    "preferred_language": "en-US",
    "system": Object {
      "codename": "n6951be2b_e502_01ed_79a4_80ad776ae16b",
      "id": "6951be2b-e502-01ed-79a4-80ad776ae16b",
      "language": "en-US",
      "last_modified": "2020-09-22T09:33:13.7465607Z",
      "name": "6951be2b-e502-01ed-79a4-80ad776ae16b",
      "sitemap_locations": Array [],
      "type": "note",
    },
  },
]
`;

exports[`webhookProcessor management API triggers upsert action via change workflow step trigger (experimental) does not call createNode actions when turned off 1`] = `Array []`;

exports[`webhookProcessor preview delivery API triggers delete operation via archive trigger call deleteNode action for modular content as well 1`] = `
Array [
  Object {
    "children": Array [],
    "elements": Object {
      "text": Object {
        "name": "Text",
        "type": "text",
        "value": "Magic note",
      },
    },
    "id": "d02ab1e7-56d4-588d-9d15-0266a84ec349",
    "internal": Object {
      "contentDigest": "aae62fcb99f01c62256c16f301df2eca",
      "counter": 145,
      "owner": "@kentico/gatsby-source-kontent",
      "type": "kontent_item_note",
    },
    "parent": null,
    "preferred_language": "en-US",
    "system": Object {
      "codename": "bcbb41f3_b964_01af_7c70_8384a101bca5",
      "id": "bcbb41f3-b964-01af-7c70-8384a101bca5",
      "language": "en-US",
      "last_modified": "2020-09-23T10:21:25.7632394Z",
      "name": "bcbb41f3-b964-01af-7c70-8384a101bca5",
      "sitemap_locations": Array [],
      "type": "note",
    },
  },
  Object {
    "children": Array [],
    "elements": Object {
      "notes": Object {
        "images": Array [],
        "links": Array [],
        "modular_content": Array [
          "test_note_linekd_inline_kontent_item",
          "bcbb41f3_b964_01af_7c70_8384a101bca5",
        ],
        "name": "Notes",
        "type": "rich_text",
        "value": "<p>Great notes</p>
<object type=\\"application/kenticocloud\\" data-type=\\"item\\" data-rel=\\"component\\" data-codename=\\"bcbb41f3_b964_01af_7c70_8384a101bca5\\"></object>
<object type=\\"application/kenticocloud\\" data-type=\\"item\\" data-rel=\\"link\\" data-codename=\\"test_note_linekd_inline_kontent_item\\"></object>",
      },
    },
    "id": "c415f1ce-6022-5672-8439-04c023ee2184",
    "internal": Object {
      "contentDigest": "f6d6bb19f908369cf2209a6de314000e",
      "counter": 144,
      "owner": "@kentico/gatsby-source-kontent",
      "type": "kontent_item_author",
    },
    "parent": null,
    "preferred_language": "en-US",
    "system": Object {
      "codename": "ondrejch_test__copy_",
      "id": "74b9fc6e-b7e2-42a1-a01b-ae6ee69b2174",
      "language": "en-US",
      "last_modified": "2020-09-23T10:21:25.7632394Z",
      "name": "Ondrejch test (copy)",
      "sitemap_locations": Array [],
      "type": "author",
    },
  },
]
`;

exports[`webhookProcessor preview delivery API triggers upsert action via upsert trigger call createNode action for modular content as well 1`] = `
Array [
  Object {
    "children": Array [],
    "elements": Object {
      "notes": Object {
        "images": Array [],
        "links": Array [],
        "modular_content": Array [
          "n6951be2b_e502_01ed_79a4_80ad776ae16b",
        ],
        "name": "Notes",
        "type": "rich_text",
        "value": "<p>This is a sample notes element with one note component</p>
<object type=\\"application/kenticocloud\\" data-type=\\"item\\" data-rel=\\"component\\" data-codename=\\"n6951be2b_e502_01ed_79a4_80ad776ae16b\\"></object>",
      },
    },
    "id": "dummyId-kontent_item_en-US_eedf0209-989c-4adb-bc83-546e325ab781",
    "internal": Object {
      "contentDigest": "4416e4041857112eddf4caeb6dca5195",
      "type": "kontent_item_author",
    },
    "preferred_language": "en-US",
    "system": Object {
      "codename": "author",
      "id": "eedf0209-989c-4adb-bc83-546e325ab781",
      "language": "en-US",
      "last_modified": "2020-09-22T09:33:13.7465607Z",
      "name": "Author",
      "sitemap_locations": Array [],
      "type": "author",
    },
  },
  Object {
    "children": Array [],
    "elements": Object {
      "text": Object {
        "name": "Text",
        "type": "text",
        "value": "Note 1",
      },
    },
    "id": "dummyId-kontent_item_en-US_6951be2b-e502-01ed-79a4-80ad776ae16b",
    "internal": Object {
      "contentDigest": "5d088a503493a36e057d990043b1515e",
      "type": "kontent_item_note",
    },
    "preferred_language": "en-US",
    "system": Object {
      "codename": "n6951be2b_e502_01ed_79a4_80ad776ae16b",
      "id": "6951be2b-e502-01ed-79a4-80ad776ae16b",
      "language": "en-US",
      "last_modified": "2020-09-22T09:33:13.7465607Z",
      "name": "6951be2b-e502-01ed-79a4-80ad776ae16b",
      "sitemap_locations": Array [],
      "type": "note",
    },
  },
]
`;
